optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,416 @@
+                                        employee, shiftList.getItems().first())
+                                                .isWeekendAndNotFirstDayOfWeekend())
+                                .penalize("startOnNotFirstDayOfWeekend", HardSoftScore.ONE_SOFT,
+                                        (employee, contract, shiftList) -> new EmployeeConsecutiveAssignmentStart(employee,
+                                                shiftList.getItems().first()).getDistanceToFirstDayOfWeekend()
+                                                * contract.getWeight());
+    }
+
+    Constraint endOnNotLastDayOfWeekend(ConstraintFactory constraintFactory) {
+        return getConsecutiveShifts(constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.COMPLETE_WEEKENDS &&
+                        minMaxContractLine.isEnabled()))
+                                .filter((employee, contract,
+                                        shiftList) -> new EmployeeConsecutiveAssignmentEnd(employee,
+                                                shiftList.getItems().last())
+                                                        .isWeekendAndNotLastDayOfWeekend())
+                                .penalize("endOnNotLastDayOfWeekend", HardSoftScore.ONE_SOFT,
+                                        (employee, contract, shiftList) -> new EmployeeConsecutiveAssignmentEnd(employee,
+                                                shiftList.getItems().last()).getDistanceToLastDayOfWeekend()
+                                                * contract.getWeight());
+    }
+
+    // Identical shiftTypes during a weekend
+    Constraint identicalShiftTypesDuringWeekend(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BooleanContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.IDENTICAL_SHIFT_TYPES_DURING_WEEKEND &&
+                        minMaxContractLine.isEnabled())
+                .join(constraintFactory.from(ShiftDate.class)
+                        .filter(date -> date.getDayOfWeek() == DayOfWeek.SUNDAY))
+                .join(constraintFactory.from(ShiftAssignment.class).filter(ShiftAssignment::isWeekend),
+                        Joiners.equal((contract, date) -> date.getWeekendSundayIndex(), ShiftAssignment::getWeekendSundayIndex),
+                        Joiners.equal((contract, date) -> contract.getContract(), ShiftAssignment::getContract))
+                .groupBy((contract, date, sa) -> contract,
+                        (contract, date, sa) -> sa.getEmployee(),
+                        (contract, date, sa) -> ImmutablePair.of(sa.getShiftType(), date), // No 4-key groupBy overload
+                        ConstraintCollectors.countTri())
```

Comment:
This is probably the only use case where using a single-key overload would have been acceptable. :-)